core: fix unrecoverable freezes of rabbit's consumer

bougue-pe opened the pull request:
Hard fix (kind of) to kill the process if a thread's exception goes up to `main()`, or if rabbit's cancel/shutdown notification is received (even if other threads may run) and let orchestrator restart it.

Bump amqp-client on the way as it doesn't hurt.

Fixes #8621
Also reproduced and fixed the case that leads to the following (different) core logs:
```
[11:16:04,880] [INFO]          [WorkerCommand] consume shutdown: amq.ctag-LsXBfmFdL6n758icthlCYA, com.rabbitmq.client.ShutdownSignalException: connection error; protocol method: #method<connection.close>(reply-code=320, reply-text=CONNECTION_FORCED - broker forced connection closure with reason 'shutdown', class-id=0, method-id=0)
[11:16:04,883] [WARN]  [ForgivingExceptionHandler] An unexpected connection driver error occurred (Exception message: Connection reset by peer)
Exception in thread "main" com.rabbitmq.client.AlreadyClosedException: connection is already closed due to connection error; protocol method: #method<connection.close>(reply-code=320, reply-text=CONNECTION_FORCED - broker forced connection closure with reason 'shutdown', class-id=0, method-id=0)
        at com.rabbitmq.client.impl.AMQConnection.startShutdown(AMQConnection.java:1012)
        at com.rabbitmq.client.impl.AMQConnection.close(AMQConnection.java:1127)
        at com.rabbitmq.client.impl.AMQConnection.close(AMQConnection.java:1056)
        at com.rabbitmq.client.impl.AMQConnection.close(AMQConnection.java:1040)
        at com.rabbitmq.client.impl.AMQConnection.close(AMQConnection.java:1032)
        at com.rabbitmq.client.impl.recovery.AutorecoveringConnection.close(AutorecoveringConnection.java:289)
        at kotlin.io.CloseableKt.closeFinally(Closeable.kt:56)
        at fr.sncf.osrd.cli.WorkerCommand.run(WorkerCommand.kt:319)
        at fr.sncf.osrd.App.main(App.java:44)
```

### Hand-tested
- [x] Run classic (one per infra) core with rabbitmq up, without editoast to load infra from: ✅core crashes.
- [x] Run single-worker core with the whole stack running: ✅core works and OSRD does its job.
- [x] Run single-worker core with the whole stack, then stop rabbitmq: ✅core exits (on shutdown notification).
- [x] Run full stack single-worker mode, then remove the `core-req-all` queue to initiate cancel notification: ✅core logs "consumer cancelled" then stops (should cover #8621).
- [x] Run full stack + single-worker editoast (no core), initiate some core requests (using front).
  - Then stop editoast and start single-worker core (so impossible to load infra inside DeliverCallback): ✅core crashes and releases unacked messages.
  - Then start editoast, then start core: ✅core works and OSRD does its job.
>  - ~Then stop editoast and start single-worker core (so impossible to load infra inside DeliverCallback): ✔️❓Exceptions in threads when trying to load infra for pending requests, core stays alive (leaving pending requests unacked until core is stopped - after step below). \
>  ➡️We can try/catch in `callback` function and `exitProcess` to force shutdown.~ DONE in last commit.
>  - ~Then start editoast (keep core as-is) and initiate some new core request (using front): ✅core "correctly" processing only new requests.~
>  - ~Stop core: ✅Unacked requests are back to ready.~
>  - ~Start core: ✅Ready requests are processed.~

### Understanding of previous and current work

Previous:
- #9591 messed-up with core crashing as expected when exception is raised up to main (still not understanding exactly what prevents the app from stopping).
- Especially #9439 was worked on (and in my memory correctly hand-tested) at the same time, but when merged (after #9591) it "silently" didn't work 😞.

Current:
- ShutdownCallback looks like it's executed in the main thread (triggering the newly added `catch` in the `main()`, then the final `return`) and it may be an idea to use shutdown notification to exit cleanly
- The thread executors are messing with the handling of exceptions
- Looks like the CancelCallback doesn't trigger a "join" or a shutodwn process (because of threads? maybe more because it's on the implementation to decide what's graceful?)

### Looks like some improvements may be done (to be explored later, sorted by ROI)
- Improve/explicit some work applied in #9439 (use `isRecoverable`, cleanup consumer logs, properly close channels and connections)
- Avoid sharing channels between threads as stated [in dedicated documentation](https://www.rabbitmq.com/client-libraries/java-api-guide#concurrency)
- Handle more standardly shutdown (on exceptions or on notification) by issuing a shutdown notification, or sharing a signal (or common variable?)

bougue-pe committed:
core: fix unrecoverable freeze on rabbit shutdown or before consuming

Hard fix (kind of) to kill the process (and let orchestrator restart) if:
* rabbit shuts down (triggering consumer's ShutdownCallback)
* or an exception before starting basicConsume() goes up to main()
  (even if other threads may run)

Bump amqp-client on the way as it doesn't hurt

Signed-off-by: Pierre-Etienne Bougué <bougue.pe@proton.me>

eckter commented on core/src/main/java/fr/sncf/osrd/cli/WorkerCommand.kt:
```diff
@@ -316,7 +320,7 @@ class WorkerCommand : CliCommand {
                 if (!channel.isOpen()) break
             }
 
-            return 0
```
How to this work out with multithreading?

I don't *quite* understand when this line is reached actually. It is when all threads have died, or just one?

Maybe some time we should rewrite this class to lower the amount of nested callbacks and functions. It's a little difficult to follow.

bougue-pe replied:
Well, not understanding it all either.

After quite a few hand-tests and trying various things I landed a different proposition: please take a look at the (updated) main comment of the PR (which may be updated again later - will notify).

bougue-pe committed:
core: force process exit on rabbit's cancel notifications

From hand-tests, shutdown is already covered by the System.exit in
App.java::main().

Signed-off-by: Pierre-Etienne Bougué <bougue.pe@proton.me>

woshilapin commented on core/src/main/java/fr/sncf/osrd/cli/WorkerCommand.kt:
```diff
@@ -303,9 +307,13 @@ class WorkerCommand : CliCommand {
-                { _ -> logger.error("consumer cancelled") },
+                { _ ->
+                    logger.error("consumer cancelled")
+                    exitProcess(0)
```
Do you still need the `exitProcess` here if in `App.java` you catch exceptions anyway?

bougue-pe replied:
This one on "cancel" is specifically required (as per my hand-tests) to really exit and address the issue from #8621.

On the other hand, after hand-tests, I'm about to remove the one on "shutdown" as the one on `App.java::main()` is reached anyway and stops the process. Looks like `ShutdownCallback` is kind of a join 😕(I will trace this in the main comment).

bougue-pe replied:
Amended the last commit: https://github.com/OpenRailAssociation/osrd/compare/3f945aea83e60c77eb5b923bf82fa292246078a4..54ad82d081978f992ddb7ed7bee04829fc9072eb
Also rephrased the 2 commits after understanding **a bit** more.

bougue-pe committed:
core: fix unrecoverable freeze on exception during rabbit's DeliverCallback

Hard fix (kind of) to kill the process (and let orchestrator restart) if an
exception goes all the way up to the DeliverCallback.
For example when not able to reach editoast for infra reload.
This will release unacked messages and move them back to ready (instead
of keeping them unacked until the worker exits).

Signed-off-by: Pierre-Etienne Bougué <bougue.pe@proton.me>

bougue-pe commented:
A third commit was pushed, and the main comment updated (and all rebased on `dev`).

No more work is planned on this, please read the main comment, any feedback is welcome, and we should be good to go 🙏

bougue-pe commented:
~There is a bit more work, actually (for me): test if it improves the case described in https://github.com/OpenRailAssociation/osrd/issues/10704~
EDIT: It does not improve things, looks like a different issue, not investigating more on it.

eckter reviewed (approved):
Thanks for handling this!

Khoyo reviewed (approved):
Great work!
